src/compiler/checker.ts

```diff
@@ -26206,7 +26206,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         if (hasOnlyExpressionInitializer(declaration) && isBlockScopedNameDeclaredBeforeUse(declaration, node)) {
             const initializer = getEffectiveInitializer(declaration);
             if (initializer) {
-                return tryGetNameFromType(getTypeOfExpression(initializer));
+                const initializerType = isBindingPattern(declaration.parent) ? getTypeForBindingElement(declaration as BindingElement) : getTypeOfExpression(initializer);

```

Comment:
I feel like I see this `isBindingPattern(declaration.parent) ? getTypeForBindingElement(declaration)` pattern over and over. I'm not sure there's a clean way to abstract it generally speaking but it seems odd to have to know to do this all over the place.

Reply:
Yeah, I just borrowed this from `getTypeForVariableLikeDeclaration`. It would be cool to not worry about this.